### PR TITLE
[FLINK-16845][connector/common] Add SourceOperator which runs the Source

### PR DESCRIPTION
**Pull request by becketqin**
## What is the purpose of the change
This PR is a part of FLIP-27.  The patch implements a `SourceOperator` which runs the `Source`. 

## Brief change log
* Introduced new interface WithOperatorCoordinator - an interface for the operators with coordinators.
* Implemented `SourceOperator` to run `Source`.
* Added unit tests and improved the `StreamTaskMailboxTestHarness`.

## Verifying this change

This change can be verified with the following tests.
org.apache.flink.streaming.api.operators.SourceOperatorTest
org.apache.flink.streaming.runtime.tasks.SourceOperatorStreamTaskTest.java

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by becketqin** [comment]
@StephanEwen Thanks for the review. I only have one concern about the second comment on `withOperatorCoordinator`.

> We need to attach the provider to the JobGraph in the end, but we can keep operator and coordinator together in the StreamTransformation and then attach it to the JobVertex from there.

It depends on whether we want to expose the operator coordinator to users or not. If we want to expose this to the users, by providing a `withOperatorCoordinator` interface, users are able to call `DataStream.transform(operator)` to leverage this mechanism. Also conceptually it seems more intuitive for the coordinator to be associated with the `Operator`, instead of a `StreamTransformation`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+			sourceReader.handleSourceEvents(((SourceEventWrapper) event).getSourceEvent());
+		} else {
+			throw new IllegalStateException("Received unexpected operator event " + event);
+		}
+	}
+
+	private void registerReader() {
+		operatorEventGateway.sendEventToCoordinator(new ReaderRegistrationEvent(
+				getRuntimeContext().getIndexOfThisSubtask(),
+				"UNKNOWN_LOCATION"));
```
[the remark this answers is not in the document]
Comment (becketqin) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:183]: Right. The location is a little tricky to provide at this point. We can probably remove it.

**Comment by becketqin** [comment]
@StephanEwen Regarding "Injecting the OperatorEventGateway via setOperatorEventGateway(...)", while I am trying to update the patch, I realized it is not possible to pass the `OperatorEventGateway` to the constructor of `SourceOperator` because the gateway is only available after register the `OperatorEventHandler` which is essentially the `SourceOperator` itself. A few solutions might be:

1. Separate out the `OperatorEventHandler` implementation from the `SourceOperator` and pass that implementation instance to the `SourceOperator` constructor. This is a little weird because the event should just be handled by the operator itself. And generally speaking, the external `OperatorEventHandler` may not be able to access some internal states of the operator.

2. Pass the `StreamOperatorParameters.containingTask()` to the constructor of `SourceOperator`. The SourceOperator will have its internal `OperatorEventHandler` class and instance created in the constructor and registered to the containing `StreamTask`. This solutions is slightly better than the previous one. But from API consistency point of view, it is less aligned because this essentially relies on the implementation of the operator to actually register an `OperatorEventHandler` in order to make everything work.

3. Add a method of `setOperatorEventGateway()` to `CoordinatedOperatorFactory` (new interface). This is essentially what `YieldingOperatorFactory` and `ProcessingTimeServiceAware` is doing. This could work, but it will make `DataStream.transform(Operator)` unusable for such operators. Instead, users have to use `DataStream.transform(OneInputStreamOperatorFactory)`. Are we going to deprecate the former interface? If so then it would make sense to put all the decorative interfaces on the factories instead of on the operators. However, in that case, the `CoordinatedOperatorFactory.createStreamOperator()` can only return a `StreamOperator`, instead of a `OperatorEventHandler`. This again makes the API loose.

Compared with the above solutions, it seems that the current `WithOperatorCoordinator` API puts together all the required pieces needed for an operator with coordinator. I kind of think API wise it is intuitive for the users to understand. As in, if an operator has a corresponding coordinator,  it should provide the coordinator as well as handle the events from the coordinator.

I am not sure what would be the best way to do it here. Do you have any suggestion?

**Comment by becketqin** [comment]
@StephanEwen I just updated the patch with the following changes.

* Combined the SplitSerializerVersion with the SplitState.
* Replaced `WithOperatorCoordinator` with `CoordinatedOperatorFactory`. The caveat is that we can only tell the implementations to return a `OperatorEventHandler` from `createStreamOperator()`.

This seems the best we can do for now. Can you take another look? Thanks!

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (StephanEwen) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java:171]: We can add the location also on the JM side, when receiving the event and forwarding it to the coordinator. I think that is how we used to do it with the previous `InputSplit` implementation.

**Comment by StephanEwen** [comment]
Thank you for looking into the comments. +1 to merge this from my side!

BTW, I was trying to apply a few changes to make things a bit simpler:
  - hide the Split <-> byte[] conversion in a specific ListState
  - consolidate the operator factories and initialize more fields eagerly
  - pull the Coordinator Provider out and pass it through the transformation.

I have the code mostly ready, will open a follow-up PR with these changes once this one is merged.

**Comment by becketqin** [comment]
@flinkbot run azure

**Comment by StephanEwen** [comment]
I created some follow-up commits, based on my suggestions above.

https://github.com/becketqin/flink/pull/3

**Comment by becketqin** [comment]
@flinkbot run azure

**Comment by becketqin** [comment]
Merged to master.
aa1ede9835192b28324789c7bbf4e6b906086797.

**Pull request closed without merging**